addons/addon-base-ui/packages/base-ui/src/parts/ForceLogout.js

```diff
@@ -0,0 +1,124 @@
+/*
+ *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License").
+ *  You may not use this file except in compliance with the License.
+ *  A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+import _ from 'lodash';
+import React from 'react';
+import { observable, action, decorate, computed, runInAction } from 'mobx';
+import { observer, inject } from 'mobx-react';
+import { Button, Modal } from 'semantic-ui-react';
+
+const jwtDecode = require('jwt-decode');
+
+// expected props
+// - authentication
+// - app
+class ForceLogout extends React.Component {
+  constructor(props) {
+    super(props);
+    runInAction(() => {
+      this.intervalId = undefined;
+    });
+  }
+
+  get app() {
+    return this.props.app;
+  }
+
+  get authentication() {
+    return this.props.authentication;
+  }
+
+  get modalOpen() {
+    return !this.state.tokenActive;
+  }
+
+  componentDidMount() {
+    this.timer = setInterval(() => {
+      const expirationInMins = this.tokenExpiresIn();
+      // eslint-disable-next-line react/no-unused-state
+      if (expirationInMins > 0) {
+        this.setState({ tokenActive: true });
```

Comment:
Any reason why you are mixing MobX state management with this.setState()?  
In general, there is no need to use this.setState() otherwise you need to worry about implementing shouldUpdate() correctly.  Using MobX makes this easier.  Instead of this.setState{{ tokenActive }).  Just declare 'tokenActive' as an observable instance variables (just like how you declared 'internvalId'.  Then use runInAction() when you need to update the tokenActive value.

Reply:
We don't actually have to `implementShouldUpdate`, right? It's usually only implemented if we want to optimize performance and don't want to rerender the REACT component on every `state` change. In our case the `state` should only change when the token has expire.

Reply:
Good point. Made the changes.

Reply:
@nguyen102  here are my thoughts:
With MobX,  it automatically provides an extremely efficient implementation of shouldComponentUpdate(), this way we never have to make a decision of when we should implement the shouldComponentUpdate().  It is better not to mix setState() if we are already using MobX to manage the state of this component.

Reply:
@hatimkhan Yeah, I agree. If we're handling local state in a class, we should use either `mobx` or `REACT state`. I personally prefer to use REACT state if we're just handling local state for a component since REACT is already built to handle that.

In terms of worrying about `shouldComponentUpdate`, from what I've read and experienced, you rarely need to implement that method. Most of the time, the default REACT implementation for `shouldComponentUpdate` is sufficient. You only need to implement that method if your component's `render` method is very large and results in long rendering time if it were to be rerender often. 

So basically the heuristic I follow is not to implement `shouldComponentUpdate` unless I see a performance impact. Seeing a performance impact on the UI is rare.

More info about this can be read [here](http://jamesknelson.com/should-i-use-shouldcomponentupdate/)

Reply:
@nguyen102 I like your heuristic, thanks for sharing the link. Since I started using MobX, I never had to think about shouldUpdateComponent at all, with MobX we never need to write one, even for data intensive/expensive components, so you completely eliminate the need for heuristic.